services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,342 @@
+          AvailabilityZone: this.region,
+        },
+        TagSpecifications: {
+          ResourceType: 'instance',
+          Tags: [{
+            Key: 'Provider',
+            Value: `wm-${this.providerId}`,
+          }, {
+            Key: 'Owner',
+            Value: workerPool.owner,
+          }, {
+            Key: 'Project',
+            Value: this.project,
+          }],
+        },
+      });
+    } catch (e) {
+      await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message, // TODO: Make sure we clear exposing this with security folks
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
+    }
+
+    Promise.all(spawned.Instances.map(i => {
+      return this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: i.InstanceId,
+        created: new Date(),
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        providerData: { // what do we want to remember, anyways? 🤔
+          project: this.project,
+        },
+      });
+    }));
+  }
+
+  async deprovision({workerPool}) {
+    // should this be implemented in Provider? Looks like it's going to be the same for all
+    await workerPool.modify(wt => {
+      wt.previousProviderIds = wt.previousProviderIds.filter(p => p !== this.providerId);
+      delete wt.providerData[this.providerId];
+    });
+    // why don't we remove workers from db or terminate instances?
+  }
+
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
+    // check worker's identity -
+    // The way AWS works, workers will be getting temporary creds for accessing AWS APIs
+    // automatically. There's no token back-and-forth
+
+    // mark it as running
+    await worker.modify(w => {
+      w.state = this.Worker.states.RUNNING;
+    });
+
+    // return object that has expires field
+    return {expires: taskcluster.fromNow('96 hours')};
+  }
+
+  async checkWorker({worker}) {
+    this.seen[worker.workerPoolId] = this.seen[worker.workerPoolId] || 0;
+
+    let instanceStatuses = (await this.ec2.describeInstanceStatus({
+      InstanceIds: [worker.workerId.toString()],
+    })).data;
+
+    Promise.all(instanceStatuses.map(is => {
+      switch (is.InstanceState.Name) {
+        case 'pending':
+        case 'running':
+          this.seen[worker.workerPoolId] += 1;
+          return Promise.resolve();
+
+        case 'shutting-down':
+        case 'terminated':
+        case 'stopping':
+        case 'stopped':
+          return worker.modify(w => {w.state = this.Worker.states.STOPPED;});
+
+        default:
+          return Promise.reject(`Unknown state: ${is.InstanceState.Name} for ${is.InstanceId}`);
+      }
+    }));
+  }
+
+  // should this be implemented on Provider? Looks like it's going to be the same for all providers
```

Comment:
@imbstack I found unanswered question^

Reply:
I think once we have a few providers in their final form, we'll find a number of good places to refactor, this being one of them -- but it's ok, and probably wise given time pressures, to wait to make those refactors later.